{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Getting Started"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's go over basic functionality and use cases of CanDI package. \n",
    "\n",
    "### Importing\n",
    "\n",
    "CanDI must be imported from from the main CanDI directory. The core CanDI objects are contained within the CanDI.candi module and are imported as follows. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import CanDI.candi as can\n",
    "#Can also be imported as \n",
    "from CanDI import candi as can"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Data Object\n",
    "The Data object is instantiated when CanDI and access as data within the candi module\n",
    "CanDI dataset paths are defined as attributes within the Data object."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/cyogodzi/projects/candi-paper/CanDI/CanDI/setup/data/depmap/CRISPR_gene_effect.csv\n",
      "/home/cyogodzi/projects/candi-paper/CanDI/CanDI/setup/data/depmap/CCLE_expression.csv\n",
      "/home/cyogodzi/projects/candi-paper/CanDI/CanDI/setup/data/depmap/CCLE_gene_cn.csv\n"
     ]
    }
   ],
   "source": [
    "print(can.data.gene_effect) # depmap ceres score\n",
    "print(can.data.expression) # ccle rna seq data\n",
    "print(can.data.gene_cn) # ccle copy number data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## How to Directly Load a Dataset\n",
    "The load method of the Data object is used to load specific datasets into memory. The datasets are saved as pandas dataframes as attributes of the data object. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ACH-001113</th>\n",
       "      <th>ACH-001289</th>\n",
       "      <th>ACH-001339</th>\n",
       "      <th>ACH-001538</th>\n",
       "      <th>ACH-000242</th>\n",
       "      <th>ACH-000708</th>\n",
       "      <th>ACH-000327</th>\n",
       "      <th>ACH-000233</th>\n",
       "      <th>ACH-000461</th>\n",
       "      <th>ACH-000705</th>\n",
       "      <th>...</th>\n",
       "      <th>ACH-000114</th>\n",
       "      <th>ACH-000402</th>\n",
       "      <th>ACH-000036</th>\n",
       "      <th>ACH-000973</th>\n",
       "      <th>ACH-001128</th>\n",
       "      <th>ACH-000750</th>\n",
       "      <th>ACH-000285</th>\n",
       "      <th>ACH-001858</th>\n",
       "      <th>ACH-001997</th>\n",
       "      <th>ACH-000052</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>TSPAN6</th>\n",
       "      <td>4.990501</td>\n",
       "      <td>5.209843</td>\n",
       "      <td>3.779260</td>\n",
       "      <td>5.726831</td>\n",
       "      <td>7.465648</td>\n",
       "      <td>4.914086</td>\n",
       "      <td>4.032982</td>\n",
       "      <td>0.097611</td>\n",
       "      <td>4.712596</td>\n",
       "      <td>5.101398</td>\n",
       "      <td>...</td>\n",
       "      <td>3.793896</td>\n",
       "      <td>0.070389</td>\n",
       "      <td>4.692650</td>\n",
       "      <td>5.026800</td>\n",
       "      <td>6.699052</td>\n",
       "      <td>4.173127</td>\n",
       "      <td>0.097611</td>\n",
       "      <td>5.045268</td>\n",
       "      <td>5.805292</td>\n",
       "      <td>4.870858</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TNMD</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.545968</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.176323</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.028569</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DPM1</th>\n",
       "      <td>7.273702</td>\n",
       "      <td>7.070604</td>\n",
       "      <td>7.346425</td>\n",
       "      <td>7.086189</td>\n",
       "      <td>6.435462</td>\n",
       "      <td>6.946848</td>\n",
       "      <td>5.806582</td>\n",
       "      <td>5.919102</td>\n",
       "      <td>6.406333</td>\n",
       "      <td>6.309976</td>\n",
       "      <td>...</td>\n",
       "      <td>6.330738</td>\n",
       "      <td>5.858230</td>\n",
       "      <td>6.623369</td>\n",
       "      <td>6.966130</td>\n",
       "      <td>6.131960</td>\n",
       "      <td>6.400879</td>\n",
       "      <td>6.428276</td>\n",
       "      <td>6.991749</td>\n",
       "      <td>7.792855</td>\n",
       "      <td>6.077457</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SCYL3</th>\n",
       "      <td>2.765535</td>\n",
       "      <td>2.538538</td>\n",
       "      <td>2.339137</td>\n",
       "      <td>2.543496</td>\n",
       "      <td>2.414136</td>\n",
       "      <td>2.577731</td>\n",
       "      <td>1.948601</td>\n",
       "      <td>3.983678</td>\n",
       "      <td>2.247928</td>\n",
       "      <td>2.361768</td>\n",
       "      <td>...</td>\n",
       "      <td>2.792855</td>\n",
       "      <td>2.757023</td>\n",
       "      <td>2.111031</td>\n",
       "      <td>1.899176</td>\n",
       "      <td>2.235727</td>\n",
       "      <td>1.807355</td>\n",
       "      <td>3.257011</td>\n",
       "      <td>1.807355</td>\n",
       "      <td>2.482848</td>\n",
       "      <td>2.304511</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C1orf112</th>\n",
       "      <td>4.480265</td>\n",
       "      <td>3.510962</td>\n",
       "      <td>4.254745</td>\n",
       "      <td>3.102658</td>\n",
       "      <td>3.864929</td>\n",
       "      <td>3.853996</td>\n",
       "      <td>2.684819</td>\n",
       "      <td>3.733354</td>\n",
       "      <td>3.032101</td>\n",
       "      <td>4.280214</td>\n",
       "      <td>...</td>\n",
       "      <td>2.643856</td>\n",
       "      <td>5.103078</td>\n",
       "      <td>2.543496</td>\n",
       "      <td>3.531069</td>\n",
       "      <td>3.971773</td>\n",
       "      <td>3.303050</td>\n",
       "      <td>4.980482</td>\n",
       "      <td>3.270529</td>\n",
       "      <td>3.903038</td>\n",
       "      <td>3.836934</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>POLR2J3</th>\n",
       "      <td>5.781884</td>\n",
       "      <td>4.704319</td>\n",
       "      <td>4.931683</td>\n",
       "      <td>3.858976</td>\n",
       "      <td>4.990501</td>\n",
       "      <td>5.303781</td>\n",
       "      <td>4.996841</td>\n",
       "      <td>6.839960</td>\n",
       "      <td>5.529196</td>\n",
       "      <td>5.860963</td>\n",
       "      <td>...</td>\n",
       "      <td>3.793896</td>\n",
       "      <td>6.669877</td>\n",
       "      <td>6.191010</td>\n",
       "      <td>5.934281</td>\n",
       "      <td>3.097611</td>\n",
       "      <td>5.102658</td>\n",
       "      <td>6.341630</td>\n",
       "      <td>4.607626</td>\n",
       "      <td>4.787119</td>\n",
       "      <td>4.452859</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>H2BE1</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.594549</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.176323</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.111031</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AL445238.1</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.028569</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.042644</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.097611</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.163499</td>\n",
       "      <td>0.163499</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GET1-SH3BGR</th>\n",
       "      <td>0.799087</td>\n",
       "      <td>0.464668</td>\n",
       "      <td>0.263034</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.263034</td>\n",
       "      <td>0.286881</td>\n",
       "      <td>2.280956</td>\n",
       "      <td>0.275007</td>\n",
       "      <td>0.790772</td>\n",
       "      <td>...</td>\n",
       "      <td>1.416840</td>\n",
       "      <td>0.526069</td>\n",
       "      <td>1.117695</td>\n",
       "      <td>0.378512</td>\n",
       "      <td>0.713696</td>\n",
       "      <td>0.214125</td>\n",
       "      <td>0.310340</td>\n",
       "      <td>1.090853</td>\n",
       "      <td>0.084064</td>\n",
       "      <td>1.422233</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC113348.1</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.070389</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.028569</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.084064</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>19177 rows × 1379 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             ACH-001113  ACH-001289  ACH-001339  ACH-001538  ACH-000242  \\\n",
       "gene                                                                      \n",
       "TSPAN6         4.990501    5.209843    3.779260    5.726831    7.465648   \n",
       "TNMD           0.000000    0.545968    0.000000    0.000000    0.000000   \n",
       "DPM1           7.273702    7.070604    7.346425    7.086189    6.435462   \n",
       "SCYL3          2.765535    2.538538    2.339137    2.543496    2.414136   \n",
       "C1orf112       4.480265    3.510962    4.254745    3.102658    3.864929   \n",
       "...                 ...         ...         ...         ...         ...   \n",
       "POLR2J3        5.781884    4.704319    4.931683    3.858976    4.990501   \n",
       "H2BE1          0.000000    0.000000    0.000000    0.000000    0.000000   \n",
       "AL445238.1     0.000000    0.000000    0.028569    0.000000    0.000000   \n",
       "GET1-SH3BGR    0.799087    0.464668    0.263034    0.000000    0.000000   \n",
       "AC113348.1     0.000000    0.070389    0.000000    0.000000    0.000000   \n",
       "\n",
       "             ACH-000708  ACH-000327  ACH-000233  ACH-000461  ACH-000705  ...  \\\n",
       "gene                                                                     ...   \n",
       "TSPAN6         4.914086    4.032982    0.097611    4.712596    5.101398  ...   \n",
       "TNMD           0.176323    0.000000    0.000000    0.000000    0.000000  ...   \n",
       "DPM1           6.946848    5.806582    5.919102    6.406333    6.309976  ...   \n",
       "SCYL3          2.577731    1.948601    3.983678    2.247928    2.361768  ...   \n",
       "C1orf112       3.853996    2.684819    3.733354    3.032101    4.280214  ...   \n",
       "...                 ...         ...         ...         ...         ...  ...   \n",
       "POLR2J3        5.303781    4.996841    6.839960    5.529196    5.860963  ...   \n",
       "H2BE1          0.000000    0.000000    0.000000    0.000000    0.594549  ...   \n",
       "AL445238.1     0.000000    0.042644    0.000000    0.000000    0.000000  ...   \n",
       "GET1-SH3BGR    0.263034    0.286881    2.280956    0.275007    0.790772  ...   \n",
       "AC113348.1     0.000000    0.028569    0.000000    0.000000    0.000000  ...   \n",
       "\n",
       "             ACH-000114  ACH-000402  ACH-000036  ACH-000973  ACH-001128  \\\n",
       "gene                                                                      \n",
       "TSPAN6         3.793896    0.070389    4.692650    5.026800    6.699052   \n",
       "TNMD           0.028569    0.000000    0.000000    0.000000    0.000000   \n",
       "DPM1           6.330738    5.858230    6.623369    6.966130    6.131960   \n",
       "SCYL3          2.792855    2.757023    2.111031    1.899176    2.235727   \n",
       "C1orf112       2.643856    5.103078    2.543496    3.531069    3.971773   \n",
       "...                 ...         ...         ...         ...         ...   \n",
       "POLR2J3        3.793896    6.669877    6.191010    5.934281    3.097611   \n",
       "H2BE1          0.000000    0.176323    0.000000    0.000000    0.000000   \n",
       "AL445238.1     0.000000    0.000000    0.000000    0.000000    0.000000   \n",
       "GET1-SH3BGR    1.416840    0.526069    1.117695    0.378512    0.713696   \n",
       "AC113348.1     0.084064    0.000000    0.000000    0.000000    0.000000   \n",
       "\n",
       "             ACH-000750  ACH-000285  ACH-001858  ACH-001997  ACH-000052  \n",
       "gene                                                                     \n",
       "TSPAN6         4.173127    0.097611    5.045268    5.805292    4.870858  \n",
       "TNMD           0.000000    0.000000    0.000000    0.000000    0.000000  \n",
       "DPM1           6.400879    6.428276    6.991749    7.792855    6.077457  \n",
       "SCYL3          1.807355    3.257011    1.807355    2.482848    2.304511  \n",
       "C1orf112       3.303050    4.980482    3.270529    3.903038    3.836934  \n",
       "...                 ...         ...         ...         ...         ...  \n",
       "POLR2J3        5.102658    6.341630    4.607626    4.787119    4.452859  \n",
       "H2BE1          0.000000    0.000000    0.111031    0.000000    0.000000  \n",
       "AL445238.1     0.097611    0.000000    0.000000    0.163499    0.163499  \n",
       "GET1-SH3BGR    0.214125    0.310340    1.090853    0.084064    1.422233  \n",
       "AC113348.1     0.000000    0.000000    0.000000    0.000000    0.000000  \n",
       "\n",
       "[19177 rows x 1379 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "can.data.load(\"expression\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Cell Lines\n",
    "The Cell Lines dataset contains all cell line metadata. This table is loaded automatically when candi is imported."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cell_line_name</th>\n",
       "      <th>stripped_cell_line_name</th>\n",
       "      <th>CCLE_Name</th>\n",
       "      <th>alias</th>\n",
       "      <th>COSMICID</th>\n",
       "      <th>sex</th>\n",
       "      <th>source</th>\n",
       "      <th>Achilles_n_replicates</th>\n",
       "      <th>cell_line_NNMD</th>\n",
       "      <th>culture_type</th>\n",
       "      <th>...</th>\n",
       "      <th>primary_or_metastasis</th>\n",
       "      <th>primary_disease</th>\n",
       "      <th>Subtype</th>\n",
       "      <th>age</th>\n",
       "      <th>Sanger_Model_ID</th>\n",
       "      <th>depmap_public_comments</th>\n",
       "      <th>lineage</th>\n",
       "      <th>lineage_subtype</th>\n",
       "      <th>lineage_sub_subtype</th>\n",
       "      <th>lineage_molecular_subtype</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DepMap_ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ACH-000001</th>\n",
       "      <td>NIH:OVCAR-3</td>\n",
       "      <td>NIHOVCAR3</td>\n",
       "      <td>NIHOVCAR3_OVARY</td>\n",
       "      <td>OVCAR3</td>\n",
       "      <td>905933.0</td>\n",
       "      <td>Female</td>\n",
       "      <td>ATCC</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>Metastasis</td>\n",
       "      <td>Ovarian Cancer</td>\n",
       "      <td>Adenocarcinoma, high grade serous</td>\n",
       "      <td>60.0</td>\n",
       "      <td>SIDM00105</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ovary</td>\n",
       "      <td>ovary_adenocarcinoma</td>\n",
       "      <td>high_grade_serous</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACH-000002</th>\n",
       "      <td>HL-60</td>\n",
       "      <td>HL60</td>\n",
       "      <td>HL60_HAEMATOPOIETIC_AND_LYMPHOID_TISSUE</td>\n",
       "      <td>NaN</td>\n",
       "      <td>905938.0</td>\n",
       "      <td>Female</td>\n",
       "      <td>ATCC</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>Primary</td>\n",
       "      <td>Leukemia</td>\n",
       "      <td>Acute Myelogenous Leukemia (AML), M3 (Promyelo...</td>\n",
       "      <td>35.0</td>\n",
       "      <td>SIDM00829</td>\n",
       "      <td>NaN</td>\n",
       "      <td>blood</td>\n",
       "      <td>AML</td>\n",
       "      <td>M3</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACH-000003</th>\n",
       "      <td>CACO2</td>\n",
       "      <td>CACO2</td>\n",
       "      <td>CACO2_LARGE_INTESTINE</td>\n",
       "      <td>CACO2, CaCo-2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Male</td>\n",
       "      <td>ATCC</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Colon/Colorectal Cancer</td>\n",
       "      <td>Adenocarcinoma</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SIDM00891</td>\n",
       "      <td>NaN</td>\n",
       "      <td>colorectal</td>\n",
       "      <td>colorectal_adenocarcinoma</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACH-000004</th>\n",
       "      <td>HEL</td>\n",
       "      <td>HEL</td>\n",
       "      <td>HEL_HAEMATOPOIETIC_AND_LYMPHOID_TISSUE</td>\n",
       "      <td>NaN</td>\n",
       "      <td>907053.0</td>\n",
       "      <td>Male</td>\n",
       "      <td>DSMZ</td>\n",
       "      <td>2.0</td>\n",
       "      <td>-3.079202</td>\n",
       "      <td>Suspension</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Leukemia</td>\n",
       "      <td>Acute Myelogenous Leukemia (AML), M6 (Erythrol...</td>\n",
       "      <td>30.0</td>\n",
       "      <td>SIDM00594</td>\n",
       "      <td>NaN</td>\n",
       "      <td>blood</td>\n",
       "      <td>AML</td>\n",
       "      <td>M6</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACH-000005</th>\n",
       "      <td>HEL 92.1.7</td>\n",
       "      <td>HEL9217</td>\n",
       "      <td>HEL9217_HAEMATOPOIETIC_AND_LYMPHOID_TISSUE</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Male</td>\n",
       "      <td>ATCC</td>\n",
       "      <td>2.0</td>\n",
       "      <td>-2.404409</td>\n",
       "      <td>Suspension</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Leukemia</td>\n",
       "      <td>Acute Myelogenous Leukemia (AML), M6 (Erythrol...</td>\n",
       "      <td>30.0</td>\n",
       "      <td>SIDM00593</td>\n",
       "      <td>NaN</td>\n",
       "      <td>blood</td>\n",
       "      <td>AML</td>\n",
       "      <td>M6</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 25 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           cell_line_name stripped_cell_line_name  \\\n",
       "DepMap_ID                                           \n",
       "ACH-000001    NIH:OVCAR-3               NIHOVCAR3   \n",
       "ACH-000002          HL-60                    HL60   \n",
       "ACH-000003          CACO2                   CACO2   \n",
       "ACH-000004            HEL                     HEL   \n",
       "ACH-000005     HEL 92.1.7                 HEL9217   \n",
       "\n",
       "                                             CCLE_Name          alias  \\\n",
       "DepMap_ID                                                               \n",
       "ACH-000001                             NIHOVCAR3_OVARY         OVCAR3   \n",
       "ACH-000002     HL60_HAEMATOPOIETIC_AND_LYMPHOID_TISSUE            NaN   \n",
       "ACH-000003                       CACO2_LARGE_INTESTINE  CACO2, CaCo-2   \n",
       "ACH-000004      HEL_HAEMATOPOIETIC_AND_LYMPHOID_TISSUE            NaN   \n",
       "ACH-000005  HEL9217_HAEMATOPOIETIC_AND_LYMPHOID_TISSUE            NaN   \n",
       "\n",
       "            COSMICID     sex source  Achilles_n_replicates  cell_line_NNMD  \\\n",
       "DepMap_ID                                                                    \n",
       "ACH-000001  905933.0  Female   ATCC                    NaN             NaN   \n",
       "ACH-000002  905938.0  Female   ATCC                    NaN             NaN   \n",
       "ACH-000003       NaN    Male   ATCC                    NaN             NaN   \n",
       "ACH-000004  907053.0    Male   DSMZ                    2.0       -3.079202   \n",
       "ACH-000005       NaN    Male   ATCC                    2.0       -2.404409   \n",
       "\n",
       "           culture_type  ... primary_or_metastasis          primary_disease  \\\n",
       "DepMap_ID                ...                                                  \n",
       "ACH-000001          NaN  ...            Metastasis           Ovarian Cancer   \n",
       "ACH-000002          NaN  ...               Primary                 Leukemia   \n",
       "ACH-000003          NaN  ...                   NaN  Colon/Colorectal Cancer   \n",
       "ACH-000004   Suspension  ...                   NaN                 Leukemia   \n",
       "ACH-000005   Suspension  ...                   NaN                 Leukemia   \n",
       "\n",
       "                                                      Subtype   age  \\\n",
       "DepMap_ID                                                             \n",
       "ACH-000001                  Adenocarcinoma, high grade serous  60.0   \n",
       "ACH-000002  Acute Myelogenous Leukemia (AML), M3 (Promyelo...  35.0   \n",
       "ACH-000003                                     Adenocarcinoma   NaN   \n",
       "ACH-000004  Acute Myelogenous Leukemia (AML), M6 (Erythrol...  30.0   \n",
       "ACH-000005  Acute Myelogenous Leukemia (AML), M6 (Erythrol...  30.0   \n",
       "\n",
       "           Sanger_Model_ID depmap_public_comments     lineage  \\\n",
       "DepMap_ID                                                       \n",
       "ACH-000001       SIDM00105                    NaN       ovary   \n",
       "ACH-000002       SIDM00829                    NaN       blood   \n",
       "ACH-000003       SIDM00891                    NaN  colorectal   \n",
       "ACH-000004       SIDM00594                    NaN       blood   \n",
       "ACH-000005       SIDM00593                    NaN       blood   \n",
       "\n",
       "                      lineage_subtype  lineage_sub_subtype  \\\n",
       "DepMap_ID                                                    \n",
       "ACH-000001       ovary_adenocarcinoma    high_grade_serous   \n",
       "ACH-000002                        AML                   M3   \n",
       "ACH-000003  colorectal_adenocarcinoma                  NaN   \n",
       "ACH-000004                        AML                   M6   \n",
       "ACH-000005                        AML                   M6   \n",
       "\n",
       "           lineage_molecular_subtype  \n",
       "DepMap_ID                             \n",
       "ACH-000001                       NaN  \n",
       "ACH-000002                       NaN  \n",
       "ACH-000003                       NaN  \n",
       "ACH-000004                       NaN  \n",
       "ACH-000005                       NaN  \n",
       "\n",
       "[5 rows x 25 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "can.data.cell_lines.head(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Genes\n",
    "The genes dataset contains relevant gene metadata. \n",
    "The genes dataset is loaded into memory automatically when candi is imported. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Approved name</th>\n",
       "      <th>Accession numbers</th>\n",
       "      <th>UniProt ID</th>\n",
       "      <th>ENTREZ ID</th>\n",
       "      <th>Ensembl ID</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Approved symbol</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>PINLYP</th>\n",
       "      <td>phospholipase A2 inhibitor and LY6/PLAUR domai...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A6NC86</td>\n",
       "      <td>390940</td>\n",
       "      <td>ENSG00000234465</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ARL6IP1P1</th>\n",
       "      <td>ADP ribosylation factor like GTPase 6 interact...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>100288702</td>\n",
       "      <td>ENSG00000255664</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PRAMEF33</th>\n",
       "      <td>PRAME family member 33</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A0A0G2JMD5</td>\n",
       "      <td>645382</td>\n",
       "      <td>ENSG00000237700</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AL353354.2</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CTA-298G8.2</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                     Approved name  \\\n",
       "Approved symbol                                                      \n",
       "PINLYP           phospholipase A2 inhibitor and LY6/PLAUR domai...   \n",
       "ARL6IP1P1        ADP ribosylation factor like GTPase 6 interact...   \n",
       "PRAMEF33                                    PRAME family member 33   \n",
       "AL353354.2                                                     NaN   \n",
       "CTA-298G8.2                                                    NaN   \n",
       "\n",
       "                Accession numbers  UniProt ID  ENTREZ ID       Ensembl ID  \n",
       "Approved symbol                                                            \n",
       "PINLYP                        NaN      A6NC86     390940  ENSG00000234465  \n",
       "ARL6IP1P1                     NaN         NaN  100288702  ENSG00000255664  \n",
       "PRAMEF33                      NaN  A0A0G2JMD5     645382  ENSG00000237700  \n",
       "AL353354.2                    NaN         NaN        NaN              NaN  \n",
       "CTA-298G8.2                   NaN         NaN        NaN              NaN  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "can.data.genes.head(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Locations\n",
    "The locations dataset contains location annotations for all genes and their associated confidence scores. Confidence scores were crowd sourced from several protein localization papers and integrated into one scale. This dataset is automatically loaded into memory when candi is imported. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>location</th>\n",
       "      <th>confidence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A1CF</td>\n",
       "      <td>Nucleus</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A4GALT</td>\n",
       "      <td>Mitochondria</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AAAS</td>\n",
       "      <td>Nucleus</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AAAS</td>\n",
       "      <td>Cytoskeleton</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AAAS</td>\n",
       "      <td>Cytosol</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     gene      location  confidence\n",
       "0    A1CF       Nucleus         1.0\n",
       "1  A4GALT  Mitochondria         1.0\n",
       "2    AAAS       Nucleus         2.0\n",
       "3    AAAS  Cytoskeleton         2.0\n",
       "4    AAAS       Cytosol         2.0"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "can.data.locations.head(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Basic Object Instantiation\n",
    "- The user input for object instantiation is used directly for indexing\n",
    "- This means if it is misspelled candi will not be able to retrieve the data in which the user is interested\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "kras = can.Gene(\"KRAS\")\n",
    "lung = can.Cancer(\"Lung Cancer\")\n",
    "membrane = can.Organelle(\"Plasma membrane\")\n",
    "a549 = can.CellLine(\"A549\") "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Gene Object Methods and Attributes\n",
    "The following function prints the internal attributes and functions of CanDI objects. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Attributes:\n",
      "\n",
      "ensembl: ENSG00000133703\n",
      "entrez: 3845\n",
      "get_name: KRAS\n",
      "name: KRAS proto-oncogene, GTPase\n",
      "symbol: KRAS\n",
      "\n",
      "Methods:\n",
      "\n",
      "cn_normal\n",
      "deletion\n",
      "dependency_of\n",
      "dependent\n",
      "duplication\n",
      "effect_of\n",
      "essential\n",
      "expressed\n",
      "expression_of\n",
      "mutated\n",
      "non_dependent\n",
      "non_essential\n",
      "unexpressed\n"
     ]
    }
   ],
   "source": [
    "def pretty_print_attr(obj):\n",
    "    attr = []\n",
    "    ls_attr = []\n",
    "    meth = []\n",
    "    for i in dir(obj):\n",
    "        if \"_\" != i[0]:\n",
    "            if type(getattr(obj, i)) == str or type(getattr(obj, i)) == int:\n",
    "                attr.append(i)\n",
    "            elif type(getattr(obj, i)) == list:\n",
    "                ls_attr.append(i)\n",
    "            else:\n",
    "                meth.append(i)\n",
    "                \n",
    "    print(\"Attributes:\\n\")\n",
    "    for i in attr: print(i+\":\", getattr(obj, i))\n",
    "    for i in ls_attr: print(i+\" list first item:\", getattr(obj, i)[0])\n",
    "    for i in ls_attr: print(i+\" length:\", len(getattr(obj, i)))\n",
    "    print(\"\\nMethods:\\n\")\n",
    "    for i in meth: print(i)\n",
    "\n",
    "pretty_print_attr(kras)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Gene Indexing examples\n",
    "If a dataset has not be loaded into memory candi will prompt you.\n",
    "Once a dataset is loaded, Gene.expression gives all the rna seq transcript data for that specific object.\n",
    "In this case we have already instantiated a gene object"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "ACH-001113    4.568640\n",
       "ACH-001289    4.554589\n",
       "ACH-001339    3.955127\n",
       "ACH-001538    5.593354\n",
       "ACH-000242    3.845992\n",
       "                ...   \n",
       "ACH-000750    3.729009\n",
       "ACH-000285    5.389567\n",
       "ACH-001858    4.014355\n",
       "ACH-001997    3.455492\n",
       "ACH-000052    3.587365\n",
       "Name: KRAS, Length: 1379, dtype: float64"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "kras.expression"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Basic CanDI filtering\n",
    "the Gene.expressed() method retrieves cell lines where the user defined gene has above 1 transcript per million\n",
    "the output is a list of cell line ids which can be used to instantiate CellLine or CellLineClbbuster objects\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['ACH-001113',\n",
       " 'ACH-001289',\n",
       " 'ACH-001339',\n",
       " 'ACH-001538',\n",
       " 'ACH-000242',\n",
       " 'ACH-000708',\n",
       " 'ACH-000327',\n",
       " 'ACH-000233',\n",
       " 'ACH-000461',\n",
       " 'ACH-000705']"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "kras.expressed()[0:10]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The user can specify if they want the tpm values with the depmap ids "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "ACH-001113    4.568640\n",
       "ACH-001289    4.554589\n",
       "ACH-001339    3.955127\n",
       "ACH-001538    5.593354\n",
       "ACH-000242    3.845992\n",
       "                ...   \n",
       "ACH-000750    3.729009\n",
       "ACH-000285    5.389567\n",
       "ACH-001858    4.014355\n",
       "ACH-001997    3.455492\n",
       "ACH-000052    3.587365\n",
       "Name: KRAS, Length: 1379, dtype: float64"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "kras.expressed(style=\"values\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "If you input a depmap id as an argument to gene.expressed you will get a boolean showing the expression status of your gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "kras.expressed(a549.depmap_id)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The user can use the gene.expression_of() method to check that gene's expression in a specific cell line.\n",
    "This method only, when called from a Gene object, accepts cell line depmap id's as an argument."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "4.350497247084133"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "kras.expression_of(a549.depmap_id)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "CanDI is consistent in the way this works across all classes and data types"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mutations has not been loaded. Do you want to load, y/n?> y\n",
      "Load Complete\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>Entrez_Gene_Id</th>\n",
       "      <th>NCBI_Build</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start_position</th>\n",
       "      <th>End_position</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Variant_Classification</th>\n",
       "      <th>Variant_Type</th>\n",
       "      <th>Reference_Allele</th>\n",
       "      <th>...</th>\n",
       "      <th>isCOSMIChotspot</th>\n",
       "      <th>COSMIChsCnt</th>\n",
       "      <th>ExAC_AF</th>\n",
       "      <th>Variant_annotation</th>\n",
       "      <th>CGA_WES_AC</th>\n",
       "      <th>HC_AC</th>\n",
       "      <th>RD_AC</th>\n",
       "      <th>RNAseq_AC</th>\n",
       "      <th>SangerWES_AC</th>\n",
       "      <th>WGS_AC</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1543</th>\n",
       "      <td>KRAS</td>\n",
       "      <td>3845</td>\n",
       "      <td>37</td>\n",
       "      <td>12</td>\n",
       "      <td>25398284</td>\n",
       "      <td>25398284</td>\n",
       "      <td>+</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>15813.0</td>\n",
       "      <td>0.000016</td>\n",
       "      <td>other non-conserving</td>\n",
       "      <td>187:172</td>\n",
       "      <td>26:35</td>\n",
       "      <td>NaN</td>\n",
       "      <td>90:89</td>\n",
       "      <td>NaN</td>\n",
       "      <td>17:12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7075</th>\n",
       "      <td>KRAS</td>\n",
       "      <td>3845</td>\n",
       "      <td>37</td>\n",
       "      <td>12</td>\n",
       "      <td>25398284</td>\n",
       "      <td>25398284</td>\n",
       "      <td>+</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>15813.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>other non-conserving</td>\n",
       "      <td>144:0</td>\n",
       "      <td>184:2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>155:2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7340</th>\n",
       "      <td>KRAS</td>\n",
       "      <td>3845</td>\n",
       "      <td>37</td>\n",
       "      <td>12</td>\n",
       "      <td>25398284</td>\n",
       "      <td>25398284</td>\n",
       "      <td>+</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>15813.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>other non-conserving</td>\n",
       "      <td>14:0</td>\n",
       "      <td>157:1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>106:1</td>\n",
       "      <td>16:0</td>\n",
       "      <td>24:0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10322</th>\n",
       "      <td>KRAS</td>\n",
       "      <td>3845</td>\n",
       "      <td>37</td>\n",
       "      <td>12</td>\n",
       "      <td>25380276</td>\n",
       "      <td>25380276</td>\n",
       "      <td>+</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>T</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>141.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>other non-conserving</td>\n",
       "      <td>34:30</td>\n",
       "      <td>97:47</td>\n",
       "      <td>NaN</td>\n",
       "      <td>52:41</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15559</th>\n",
       "      <td>KRAS</td>\n",
       "      <td>3845</td>\n",
       "      <td>37</td>\n",
       "      <td>12</td>\n",
       "      <td>25398284</td>\n",
       "      <td>25398284</td>\n",
       "      <td>+</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>15813.0</td>\n",
       "      <td>0.000016</td>\n",
       "      <td>other non-conserving</td>\n",
       "      <td>14:20</td>\n",
       "      <td>39:45</td>\n",
       "      <td>NaN</td>\n",
       "      <td>91:89</td>\n",
       "      <td>23:30</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1265558</th>\n",
       "      <td>KRAS</td>\n",
       "      <td>3845</td>\n",
       "      <td>37</td>\n",
       "      <td>12</td>\n",
       "      <td>25398283</td>\n",
       "      <td>25398284</td>\n",
       "      <td>+</td>\n",
       "      <td>In_Frame_Ins</td>\n",
       "      <td>INS</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>15827.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>other non-conserving</td>\n",
       "      <td>71:112</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1265728</th>\n",
       "      <td>KRAS</td>\n",
       "      <td>3845</td>\n",
       "      <td>37</td>\n",
       "      <td>12</td>\n",
       "      <td>25378562</td>\n",
       "      <td>25378562</td>\n",
       "      <td>+</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>82.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>other non-conserving</td>\n",
       "      <td>58:71</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1265729</th>\n",
       "      <td>KRAS</td>\n",
       "      <td>3845</td>\n",
       "      <td>37</td>\n",
       "      <td>12</td>\n",
       "      <td>25398283</td>\n",
       "      <td>25398284</td>\n",
       "      <td>+</td>\n",
       "      <td>In_Frame_Ins</td>\n",
       "      <td>INS</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>15827.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>other non-conserving</td>\n",
       "      <td>76:106</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1265899</th>\n",
       "      <td>KRAS</td>\n",
       "      <td>3845</td>\n",
       "      <td>37</td>\n",
       "      <td>12</td>\n",
       "      <td>25398283</td>\n",
       "      <td>25398284</td>\n",
       "      <td>+</td>\n",
       "      <td>In_Frame_Ins</td>\n",
       "      <td>INS</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>15827.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>other non-conserving</td>\n",
       "      <td>55:70</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1266065</th>\n",
       "      <td>KRAS</td>\n",
       "      <td>3845</td>\n",
       "      <td>37</td>\n",
       "      <td>12</td>\n",
       "      <td>25398283</td>\n",
       "      <td>25398284</td>\n",
       "      <td>+</td>\n",
       "      <td>In_Frame_Ins</td>\n",
       "      <td>INS</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>15827.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>other non-conserving</td>\n",
       "      <td>67:78</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>285 rows × 32 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         gene  Entrez_Gene_Id  NCBI_Build Chromosome  Start_position  \\\n",
       "1543     KRAS            3845          37         12        25398284   \n",
       "7075     KRAS            3845          37         12        25398284   \n",
       "7340     KRAS            3845          37         12        25398284   \n",
       "10322    KRAS            3845          37         12        25380276   \n",
       "15559    KRAS            3845          37         12        25398284   \n",
       "...       ...             ...         ...        ...             ...   \n",
       "1265558  KRAS            3845          37         12        25398283   \n",
       "1265728  KRAS            3845          37         12        25378562   \n",
       "1265729  KRAS            3845          37         12        25398283   \n",
       "1265899  KRAS            3845          37         12        25398283   \n",
       "1266065  KRAS            3845          37         12        25398283   \n",
       "\n",
       "         End_position Strand Variant_Classification Variant_Type  \\\n",
       "1543         25398284      +      Missense_Mutation          SNP   \n",
       "7075         25398284      +      Missense_Mutation          SNP   \n",
       "7340         25398284      +      Missense_Mutation          SNP   \n",
       "10322        25380276      +      Missense_Mutation          SNP   \n",
       "15559        25398284      +      Missense_Mutation          SNP   \n",
       "...               ...    ...                    ...          ...   \n",
       "1265558      25398284      +           In_Frame_Ins          INS   \n",
       "1265728      25378562      +      Missense_Mutation          SNP   \n",
       "1265729      25398284      +           In_Frame_Ins          INS   \n",
       "1265899      25398284      +           In_Frame_Ins          INS   \n",
       "1266065      25398284      +           In_Frame_Ins          INS   \n",
       "\n",
       "        Reference_Allele  ... isCOSMIChotspot COSMIChsCnt   ExAC_AF  \\\n",
       "1543                   C  ...            True     15813.0  0.000016   \n",
       "7075                   C  ...            True     15813.0       NaN   \n",
       "7340                   C  ...            True     15813.0       NaN   \n",
       "10322                  T  ...            True       141.0       NaN   \n",
       "15559                  C  ...            True     15813.0  0.000016   \n",
       "...                  ...  ...             ...         ...       ...   \n",
       "1265558                -  ...            True     15827.0       NaN   \n",
       "1265728                C  ...            True        82.0       NaN   \n",
       "1265729                -  ...            True     15827.0       NaN   \n",
       "1265899                -  ...            True     15827.0       NaN   \n",
       "1266065                -  ...            True     15827.0       NaN   \n",
       "\n",
       "           Variant_annotation CGA_WES_AC  HC_AC RD_AC RNAseq_AC SangerWES_AC  \\\n",
       "1543     other non-conserving    187:172  26:35   NaN     90:89          NaN   \n",
       "7075     other non-conserving      144:0  184:2   NaN     155:2          NaN   \n",
       "7340     other non-conserving       14:0  157:1   NaN     106:1         16:0   \n",
       "10322    other non-conserving      34:30  97:47   NaN     52:41          NaN   \n",
       "15559    other non-conserving      14:20  39:45   NaN     91:89        23:30   \n",
       "...                       ...        ...    ...   ...       ...          ...   \n",
       "1265558  other non-conserving     71:112    NaN   NaN       NaN          NaN   \n",
       "1265728  other non-conserving      58:71    NaN   NaN       NaN          NaN   \n",
       "1265729  other non-conserving     76:106    NaN   NaN       NaN          NaN   \n",
       "1265899  other non-conserving      55:70    NaN   NaN       NaN          NaN   \n",
       "1266065  other non-conserving      67:78    NaN   NaN       NaN          NaN   \n",
       "\n",
       "        WGS_AC  \n",
       "1543     17:12  \n",
       "7075       NaN  \n",
       "7340      24:0  \n",
       "10322      NaN  \n",
       "15559      NaN  \n",
       "...        ...  \n",
       "1265558    NaN  \n",
       "1265728    NaN  \n",
       "1265729    NaN  \n",
       "1265899    NaN  \n",
       "1266065    NaN  \n",
       "\n",
       "[285 rows x 32 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "kras.mutations"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The gene.mutated() method allows very specific filtering.\n",
    "Using the variant argument one can select the column on which to filter. Then using the item argument the user can specifiy the specific value in which they're interested. The example below shows retrieval of all cell lines with kras missense mutations."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['ACH-000094',\n",
       " 'ACH-000178',\n",
       " 'ACH-002186',\n",
       " 'ACH-000311',\n",
       " 'ACH-001345',\n",
       " 'ACH-001843',\n",
       " 'ACH-001353',\n",
       " 'ACH-000417',\n",
       " 'ACH-000347',\n",
       " 'ACH-000997']"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "kras.mutated(variant=\"Variant_Classification\", item=\"Missense_Mutation\")[0:10]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Users can use the unload method of the Data object to remove a dataset from memory and return it to a file path string."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "PosixPath('/home/cyogodzi/projects/candi-paper/CanDI/CanDI/setup/data/depmap/CCLE_mutations.csv')"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "can.data.unload('mutations')\n",
    "can.data.mutations"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## CellLine Methods and Attributes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Attributes:\n",
      "\n",
      "ccle_name: A549_LUNG\n",
      "depmap_id: ACH-000681\n",
      "get_name: ACH-000681\n",
      "lineage: lung\n",
      "name: A549\n",
      "sanger_id: SIDM00903\n",
      "sex: Male\n",
      "source: ATCC\n",
      "subtype: NSCLC\n",
      "tissue: lung\n",
      "\n",
      "Methods:\n",
      "\n",
      "aliases\n",
      "cn_normal\n",
      "cosmic_id\n",
      "deletion\n",
      "dependency_of\n",
      "dependent\n",
      "duplication\n",
      "effect_of\n",
      "essential\n",
      "expressed\n",
      "expression_of\n",
      "mutated\n",
      "non_dependent\n",
      "non_essential\n",
      "unexpressed\n"
     ]
    }
   ],
   "source": [
    "pretty_print_attr(a549)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "All methods work in essentially same way regardless of the candi object in use.\n",
    "The CellLine.expressed() method will return all genes which have expression above 1 transcript per million\n",
    "in that specific cell line."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['TSPAN6',\n",
       " 'DPM1',\n",
       " 'SCYL3',\n",
       " 'C1orf112',\n",
       " 'CFH',\n",
       " 'FUCA2',\n",
       " 'GCLC',\n",
       " 'NFYA',\n",
       " 'STPG1',\n",
       " 'NIPAL3']"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "a549.expressed()[:10]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Just like gene.expressed() the user can ask for the values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "gene\n",
       "TSPAN6         5.176323\n",
       "DPM1           6.310522\n",
       "SCYL3          2.017922\n",
       "C1orf112       4.058316\n",
       "CFH            3.772941\n",
       "                 ...   \n",
       "UPK3BL2        1.367371\n",
       "AC093512.2     4.087463\n",
       "ARHGAP11B      1.531069\n",
       "ABCF2-H2BE1    1.891419\n",
       "POLR2J3        3.372952\n",
       "Name: ACH-000681, Length: 11498, dtype: float64"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "a549.expressed(style=\"values\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "And for specific genes expression status"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "a549.expressed(\"KRAS\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "expressed with style=\"values\" gives the same result as expression_of"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "4.350497247084133"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "a549.expression_of(\"KRAS\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "4.350497247084133"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "a549.expressed(\"KRAS\", style=\"values\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The CellLine.mtuations attribute gives all mutation data for that specific cell line"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mutations has not been loaded. Do you want to load, y/n?> y\n",
      "Load Complete\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>Entrez_Gene_Id</th>\n",
       "      <th>NCBI_Build</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start_position</th>\n",
       "      <th>End_position</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Variant_Classification</th>\n",
       "      <th>Variant_Type</th>\n",
       "      <th>Reference_Allele</th>\n",
       "      <th>...</th>\n",
       "      <th>isCOSMIChotspot</th>\n",
       "      <th>COSMIChsCnt</th>\n",
       "      <th>ExAC_AF</th>\n",
       "      <th>Variant_annotation</th>\n",
       "      <th>CGA_WES_AC</th>\n",
       "      <th>HC_AC</th>\n",
       "      <th>RD_AC</th>\n",
       "      <th>RNAseq_AC</th>\n",
       "      <th>SangerWES_AC</th>\n",
       "      <th>WGS_AC</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>244692</th>\n",
       "      <td>TPRG1L</td>\n",
       "      <td>127262</td>\n",
       "      <td>37</td>\n",
       "      <td>1</td>\n",
       "      <td>3542384</td>\n",
       "      <td>3542384</td>\n",
       "      <td>+</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>other non-conserving</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>17:28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>244693</th>\n",
       "      <td>ENO1</td>\n",
       "      <td>2023</td>\n",
       "      <td>37</td>\n",
       "      <td>1</td>\n",
       "      <td>8925414</td>\n",
       "      <td>8925414</td>\n",
       "      <td>+</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>other non-conserving</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>22:30</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>244694</th>\n",
       "      <td>NMNAT1</td>\n",
       "      <td>64802</td>\n",
       "      <td>37</td>\n",
       "      <td>1</td>\n",
       "      <td>10042579</td>\n",
       "      <td>10042579</td>\n",
       "      <td>+</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>other non-conserving</td>\n",
       "      <td>33:30</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>13:33</td>\n",
       "      <td>33:31</td>\n",
       "      <td>20:32</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>244695</th>\n",
       "      <td>MFN2</td>\n",
       "      <td>9927</td>\n",
       "      <td>37</td>\n",
       "      <td>1</td>\n",
       "      <td>12058908</td>\n",
       "      <td>12058908</td>\n",
       "      <td>+</td>\n",
       "      <td>Silent</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>silent</td>\n",
       "      <td>19:91</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>20:93</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>244696</th>\n",
       "      <td>PRAMEF4</td>\n",
       "      <td>400735</td>\n",
       "      <td>37</td>\n",
       "      <td>1</td>\n",
       "      <td>12942971</td>\n",
       "      <td>12942971</td>\n",
       "      <td>+</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>other non-conserving</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>29:39</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>245445</th>\n",
       "      <td>IGSF1</td>\n",
       "      <td>3547</td>\n",
       "      <td>37</td>\n",
       "      <td>X</td>\n",
       "      <td>130411178</td>\n",
       "      <td>130411178</td>\n",
       "      <td>+</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>other non-conserving</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>19:12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>245446</th>\n",
       "      <td>HS6ST2</td>\n",
       "      <td>90161</td>\n",
       "      <td>37</td>\n",
       "      <td>X</td>\n",
       "      <td>132091282</td>\n",
       "      <td>132091282</td>\n",
       "      <td>+</td>\n",
       "      <td>Silent</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>silent</td>\n",
       "      <td>35:30</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>35:32</td>\n",
       "      <td>16:10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>245447</th>\n",
       "      <td>SLITRK4</td>\n",
       "      <td>139065</td>\n",
       "      <td>37</td>\n",
       "      <td>X</td>\n",
       "      <td>142717709</td>\n",
       "      <td>142717709</td>\n",
       "      <td>+</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>other non-conserving</td>\n",
       "      <td>125:0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>128:0</td>\n",
       "      <td>37:0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>245448</th>\n",
       "      <td>MAGEA11</td>\n",
       "      <td>4110</td>\n",
       "      <td>37</td>\n",
       "      <td>X</td>\n",
       "      <td>148798368</td>\n",
       "      <td>148798368</td>\n",
       "      <td>+</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>G</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>other non-conserving</td>\n",
       "      <td>96:1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>69:1</td>\n",
       "      <td>47:0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>245449</th>\n",
       "      <td>MAMLD1</td>\n",
       "      <td>10046</td>\n",
       "      <td>37</td>\n",
       "      <td>X</td>\n",
       "      <td>149639149</td>\n",
       "      <td>149639149</td>\n",
       "      <td>+</td>\n",
       "      <td>Missense_Mutation</td>\n",
       "      <td>SNP</td>\n",
       "      <td>C</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>other non-conserving</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>14:26</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>758 rows × 32 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           gene  Entrez_Gene_Id  NCBI_Build Chromosome  Start_position  \\\n",
       "244692   TPRG1L          127262          37          1         3542384   \n",
       "244693     ENO1            2023          37          1         8925414   \n",
       "244694   NMNAT1           64802          37          1        10042579   \n",
       "244695     MFN2            9927          37          1        12058908   \n",
       "244696  PRAMEF4          400735          37          1        12942971   \n",
       "...         ...             ...         ...        ...             ...   \n",
       "245445    IGSF1            3547          37          X       130411178   \n",
       "245446   HS6ST2           90161          37          X       132091282   \n",
       "245447  SLITRK4          139065          37          X       142717709   \n",
       "245448  MAGEA11            4110          37          X       148798368   \n",
       "245449   MAMLD1           10046          37          X       149639149   \n",
       "\n",
       "        End_position Strand Variant_Classification Variant_Type  \\\n",
       "244692       3542384      +      Missense_Mutation          SNP   \n",
       "244693       8925414      +      Missense_Mutation          SNP   \n",
       "244694      10042579      +      Missense_Mutation          SNP   \n",
       "244695      12058908      +                 Silent          SNP   \n",
       "244696      12942971      +      Missense_Mutation          SNP   \n",
       "...              ...    ...                    ...          ...   \n",
       "245445     130411178      +      Missense_Mutation          SNP   \n",
       "245446     132091282      +                 Silent          SNP   \n",
       "245447     142717709      +      Missense_Mutation          SNP   \n",
       "245448     148798368      +      Missense_Mutation          SNP   \n",
       "245449     149639149      +      Missense_Mutation          SNP   \n",
       "\n",
       "       Reference_Allele  ... isCOSMIChotspot COSMIChsCnt ExAC_AF  \\\n",
       "244692                G  ...           False         0.0     NaN   \n",
       "244693                A  ...           False         0.0     NaN   \n",
       "244694                C  ...           False         0.0     NaN   \n",
       "244695                C  ...           False         0.0     NaN   \n",
       "244696                G  ...           False         0.0     NaN   \n",
       "...                 ...  ...             ...         ...     ...   \n",
       "245445                G  ...           False         0.0     NaN   \n",
       "245446                G  ...           False         0.0     NaN   \n",
       "245447                G  ...           False         0.0     NaN   \n",
       "245448                G  ...           False         0.0     NaN   \n",
       "245449                C  ...           False         0.0     NaN   \n",
       "\n",
       "          Variant_annotation CGA_WES_AC HC_AC RD_AC RNAseq_AC SangerWES_AC  \\\n",
       "244692  other non-conserving        NaN   NaN   NaN       NaN          NaN   \n",
       "244693  other non-conserving        NaN   NaN   NaN       NaN          NaN   \n",
       "244694  other non-conserving      33:30   NaN   NaN     13:33        33:31   \n",
       "244695                silent      19:91   NaN   NaN       NaN        20:93   \n",
       "244696  other non-conserving        NaN   NaN   NaN       NaN          NaN   \n",
       "...                      ...        ...   ...   ...       ...          ...   \n",
       "245445  other non-conserving        NaN   NaN   NaN       NaN          NaN   \n",
       "245446                silent      35:30   NaN   NaN       NaN        35:32   \n",
       "245447  other non-conserving      125:0   NaN   NaN       NaN        128:0   \n",
       "245448  other non-conserving       96:1   NaN   NaN       NaN         69:1   \n",
       "245449  other non-conserving        NaN   NaN   NaN       NaN          NaN   \n",
       "\n",
       "       WGS_AC  \n",
       "244692  17:28  \n",
       "244693  22:30  \n",
       "244694  20:32  \n",
       "244695    NaN  \n",
       "244696  29:39  \n",
       "...       ...  \n",
       "245445  19:12  \n",
       "245446  16:10  \n",
       "245447   37:0  \n",
       "245448   47:0  \n",
       "245449  14:26  \n",
       "\n",
       "[758 rows x 32 columns]"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "a549.mutations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# calling the CellLine.mutated() method works the same way with all CanDI objects\n",
    "a549.mutated(variant=\"Variant_Classification\", item=\"Nonsense_Mutation\")[:10]\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Cancer Methods and Attributes\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Attributes:\n",
      "\n",
      "disease: Lung Cancer\n",
      "get_name: Lung Cancer\n",
      "ccle_names list first item: NCIH2077_LUNG\n",
      "depmap_ids list first item: ACH-000010\n",
      "names list first item: NCI-H2077\n",
      "ccle_names length: 273\n",
      "depmap_ids length: 273\n",
      "names length: 273\n",
      "\n",
      "Methods:\n",
      "\n",
      "cn_normal\n",
      "deletion\n",
      "dependency_of\n",
      "dependent\n",
      "duplication\n",
      "effect_of\n",
      "essential\n",
      "expressed\n",
      "expression_of\n",
      "mutated\n",
      "mutation_matrix\n",
      "non_dependent\n",
      "non_essential\n",
      "sexes\n",
      "sources\n",
      "subtypes\n",
      "unexpressed\n"
     ]
    }
   ],
   "source": [
    "pretty_print_attr(lung)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Cancer objects work essentially works as a group of cell line objects \n",
    "the Cancer.expression object returns a pandas dataframe rather than a pandas series since there are multiple cell lines to consider."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ACH-000010</th>\n",
       "      <th>ACH-000012</th>\n",
       "      <th>ACH-000015</th>\n",
       "      <th>ACH-000021</th>\n",
       "      <th>ACH-000029</th>\n",
       "      <th>ACH-000030</th>\n",
       "      <th>ACH-000033</th>\n",
       "      <th>ACH-000035</th>\n",
       "      <th>ACH-000062</th>\n",
       "      <th>ACH-000066</th>\n",
       "      <th>...</th>\n",
       "      <th>ACH-001386</th>\n",
       "      <th>ACH-001549</th>\n",
       "      <th>ACH-001555</th>\n",
       "      <th>ACH-001556</th>\n",
       "      <th>ACH-001557</th>\n",
       "      <th>ACH-001558</th>\n",
       "      <th>ACH-001559</th>\n",
       "      <th>ACH-001560</th>\n",
       "      <th>ACH-001561</th>\n",
       "      <th>ACH-001562</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>TSPAN6</th>\n",
       "      <td>4.447579</td>\n",
       "      <td>5.802452</td>\n",
       "      <td>4.794936</td>\n",
       "      <td>4.831371</td>\n",
       "      <td>6.491532</td>\n",
       "      <td>5.399855</td>\n",
       "      <td>5.391974</td>\n",
       "      <td>4.888013</td>\n",
       "      <td>4.731726</td>\n",
       "      <td>5.753818</td>\n",
       "      <td>...</td>\n",
       "      <td>3.060047</td>\n",
       "      <td>0.773996</td>\n",
       "      <td>2.769772</td>\n",
       "      <td>4.280956</td>\n",
       "      <td>4.089159</td>\n",
       "      <td>4.799087</td>\n",
       "      <td>4.275007</td>\n",
       "      <td>4.621173</td>\n",
       "      <td>4.399855</td>\n",
       "      <td>4.628774</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TNMD</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.028569</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.028569</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.028569</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.028569</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DPM1</th>\n",
       "      <td>7.227760</td>\n",
       "      <td>5.997744</td>\n",
       "      <td>6.929436</td>\n",
       "      <td>6.498570</td>\n",
       "      <td>6.672850</td>\n",
       "      <td>6.307246</td>\n",
       "      <td>6.866661</td>\n",
       "      <td>7.254840</td>\n",
       "      <td>5.849499</td>\n",
       "      <td>5.514438</td>\n",
       "      <td>...</td>\n",
       "      <td>6.837691</td>\n",
       "      <td>5.908333</td>\n",
       "      <td>7.010220</td>\n",
       "      <td>7.745170</td>\n",
       "      <td>7.149137</td>\n",
       "      <td>7.057992</td>\n",
       "      <td>7.379725</td>\n",
       "      <td>7.168923</td>\n",
       "      <td>7.191109</td>\n",
       "      <td>7.704941</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SCYL3</th>\n",
       "      <td>2.405992</td>\n",
       "      <td>1.970854</td>\n",
       "      <td>2.952334</td>\n",
       "      <td>2.414136</td>\n",
       "      <td>2.475085</td>\n",
       "      <td>2.017922</td>\n",
       "      <td>1.803227</td>\n",
       "      <td>2.289834</td>\n",
       "      <td>2.503349</td>\n",
       "      <td>2.536053</td>\n",
       "      <td>...</td>\n",
       "      <td>2.472488</td>\n",
       "      <td>1.831877</td>\n",
       "      <td>1.895303</td>\n",
       "      <td>2.304511</td>\n",
       "      <td>2.220330</td>\n",
       "      <td>2.204767</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>1.871844</td>\n",
       "      <td>2.060047</td>\n",
       "      <td>2.589763</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C1orf112</th>\n",
       "      <td>4.594549</td>\n",
       "      <td>3.784504</td>\n",
       "      <td>3.709291</td>\n",
       "      <td>4.527946</td>\n",
       "      <td>4.486714</td>\n",
       "      <td>3.671293</td>\n",
       "      <td>3.841973</td>\n",
       "      <td>3.795975</td>\n",
       "      <td>3.761285</td>\n",
       "      <td>4.566206</td>\n",
       "      <td>...</td>\n",
       "      <td>4.571677</td>\n",
       "      <td>4.060912</td>\n",
       "      <td>1.550901</td>\n",
       "      <td>4.306700</td>\n",
       "      <td>4.551516</td>\n",
       "      <td>3.526069</td>\n",
       "      <td>4.251719</td>\n",
       "      <td>2.153805</td>\n",
       "      <td>3.231125</td>\n",
       "      <td>3.066950</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>POLR2J3</th>\n",
       "      <td>4.516646</td>\n",
       "      <td>5.238023</td>\n",
       "      <td>5.458119</td>\n",
       "      <td>6.648465</td>\n",
       "      <td>5.239933</td>\n",
       "      <td>5.650190</td>\n",
       "      <td>6.129489</td>\n",
       "      <td>4.511595</td>\n",
       "      <td>5.559798</td>\n",
       "      <td>5.768449</td>\n",
       "      <td>...</td>\n",
       "      <td>5.689020</td>\n",
       "      <td>3.447579</td>\n",
       "      <td>5.375735</td>\n",
       "      <td>3.456806</td>\n",
       "      <td>4.397118</td>\n",
       "      <td>5.673556</td>\n",
       "      <td>3.759156</td>\n",
       "      <td>4.130931</td>\n",
       "      <td>4.688740</td>\n",
       "      <td>5.414474</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>H2BE1</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AL445238.1</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.124328</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.298658</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GET1-SH3BGR</th>\n",
       "      <td>0.807355</td>\n",
       "      <td>0.536053</td>\n",
       "      <td>0.545968</td>\n",
       "      <td>0.757023</td>\n",
       "      <td>0.150560</td>\n",
       "      <td>0.650765</td>\n",
       "      <td>0.941106</td>\n",
       "      <td>0.432959</td>\n",
       "      <td>0.748461</td>\n",
       "      <td>0.536053</td>\n",
       "      <td>...</td>\n",
       "      <td>0.321928</td>\n",
       "      <td>0.739848</td>\n",
       "      <td>0.739848</td>\n",
       "      <td>0.495695</td>\n",
       "      <td>1.659925</td>\n",
       "      <td>1.028569</td>\n",
       "      <td>0.389567</td>\n",
       "      <td>1.327687</td>\n",
       "      <td>0.650765</td>\n",
       "      <td>0.807355</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC113348.1</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.042644</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.042644</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.432959</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.189034</td>\n",
       "      <td>0.042644</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>19177 rows × 206 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             ACH-000010  ACH-000012  ACH-000015  ACH-000021  ACH-000029  \\\n",
       "gene                                                                      \n",
       "TSPAN6         4.447579    5.802452    4.794936    4.831371    6.491532   \n",
       "TNMD           0.000000    0.000000    0.000000    0.000000    0.000000   \n",
       "DPM1           7.227760    5.997744    6.929436    6.498570    6.672850   \n",
       "SCYL3          2.405992    1.970854    2.952334    2.414136    2.475085   \n",
       "C1orf112       4.594549    3.784504    3.709291    4.527946    4.486714   \n",
       "...                 ...         ...         ...         ...         ...   \n",
       "POLR2J3        4.516646    5.238023    5.458119    6.648465    5.239933   \n",
       "H2BE1          0.000000    0.000000    0.000000    0.000000    0.000000   \n",
       "AL445238.1     0.000000    0.000000    0.000000    0.000000    0.000000   \n",
       "GET1-SH3BGR    0.807355    0.536053    0.545968    0.757023    0.150560   \n",
       "AC113348.1     0.000000    0.000000    0.000000    0.000000    0.000000   \n",
       "\n",
       "             ACH-000030  ACH-000033  ACH-000035  ACH-000062  ACH-000066  ...  \\\n",
       "gene                                                                     ...   \n",
       "TSPAN6         5.399855    5.391974    4.888013    4.731726    5.753818  ...   \n",
       "TNMD           0.000000    0.000000    0.028569    0.000000    0.000000  ...   \n",
       "DPM1           6.307246    6.866661    7.254840    5.849499    5.514438  ...   \n",
       "SCYL3          2.017922    1.803227    2.289834    2.503349    2.536053  ...   \n",
       "C1orf112       3.671293    3.841973    3.795975    3.761285    4.566206  ...   \n",
       "...                 ...         ...         ...         ...         ...  ...   \n",
       "POLR2J3        5.650190    6.129489    4.511595    5.559798    5.768449  ...   \n",
       "H2BE1          0.000000    0.000000    0.000000    0.000000    0.000000  ...   \n",
       "AL445238.1     0.124328    0.000000    0.000000    0.298658    0.000000  ...   \n",
       "GET1-SH3BGR    0.650765    0.941106    0.432959    0.748461    0.536053  ...   \n",
       "AC113348.1     0.000000    0.000000    0.042644    0.000000    0.000000  ...   \n",
       "\n",
       "             ACH-001386  ACH-001549  ACH-001555  ACH-001556  ACH-001557  \\\n",
       "gene                                                                      \n",
       "TSPAN6         3.060047    0.773996    2.769772    4.280956    4.089159   \n",
       "TNMD           0.000000    0.028569    0.000000    0.000000    0.028569   \n",
       "DPM1           6.837691    5.908333    7.010220    7.745170    7.149137   \n",
       "SCYL3          2.472488    1.831877    1.895303    2.304511    2.220330   \n",
       "C1orf112       4.571677    4.060912    1.550901    4.306700    4.551516   \n",
       "...                 ...         ...         ...         ...         ...   \n",
       "POLR2J3        5.689020    3.447579    5.375735    3.456806    4.397118   \n",
       "H2BE1          0.000000    0.000000    0.000000    0.000000    0.000000   \n",
       "AL445238.1     0.000000    0.000000    0.000000    0.000000    0.000000   \n",
       "GET1-SH3BGR    0.321928    0.739848    0.739848    0.495695    1.659925   \n",
       "AC113348.1     0.042644    0.000000    0.000000    0.000000    0.000000   \n",
       "\n",
       "             ACH-001558  ACH-001559  ACH-001560  ACH-001561  ACH-001562  \n",
       "gene                                                                     \n",
       "TSPAN6         4.799087    4.275007    4.621173    4.399855    4.628774  \n",
       "TNMD           0.000000    0.000000    0.000000    0.028569    0.000000  \n",
       "DPM1           7.057992    7.379725    7.168923    7.191109    7.704941  \n",
       "SCYL3          2.204767    2.000000    1.871844    2.060047    2.589763  \n",
       "C1orf112       3.526069    4.251719    2.153805    3.231125    3.066950  \n",
       "...                 ...         ...         ...         ...         ...  \n",
       "POLR2J3        5.673556    3.759156    4.130931    4.688740    5.414474  \n",
       "H2BE1          0.000000    0.000000    0.000000    0.000000    0.000000  \n",
       "AL445238.1     0.000000    0.000000    0.000000    0.000000    0.000000  \n",
       "GET1-SH3BGR    1.028569    0.389567    1.327687    0.650765    0.807355  \n",
       "AC113348.1     0.432959    0.000000    0.000000    0.189034    0.042644  \n",
       "\n",
       "[19177 rows x 206 columns]"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "lung.expression"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Cancer.expressed method uses an abitrary threshold to filter genes the default is if a gene is expressed in 100 percent of the cell lines within the cancer object it will read out as expressed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['DPM1',\n",
       " 'SCYL3',\n",
       " 'C1orf112',\n",
       " 'GCLC',\n",
       " 'NFYA',\n",
       " 'LAS1L',\n",
       " 'ANKIB1',\n",
       " 'CYP51A1',\n",
       " 'KRIT1',\n",
       " 'RAD52']"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "lung.expressed()[0:10]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The user can relax this threshold as necessary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['TSPAN6',\n",
       " 'DPM1',\n",
       " 'SCYL3',\n",
       " 'C1orf112',\n",
       " 'CFH',\n",
       " 'FUCA2',\n",
       " 'GCLC',\n",
       " 'NFYA',\n",
       " 'STPG1',\n",
       " 'NIPAL3']"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "lung.expressed(threshold=0.50)[0:10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ACH-000010</th>\n",
       "      <th>ACH-000012</th>\n",
       "      <th>ACH-000015</th>\n",
       "      <th>ACH-000021</th>\n",
       "      <th>ACH-000029</th>\n",
       "      <th>ACH-000030</th>\n",
       "      <th>ACH-000033</th>\n",
       "      <th>ACH-000035</th>\n",
       "      <th>ACH-000062</th>\n",
       "      <th>ACH-000066</th>\n",
       "      <th>...</th>\n",
       "      <th>ACH-001386</th>\n",
       "      <th>ACH-001549</th>\n",
       "      <th>ACH-001555</th>\n",
       "      <th>ACH-001556</th>\n",
       "      <th>ACH-001557</th>\n",
       "      <th>ACH-001558</th>\n",
       "      <th>ACH-001559</th>\n",
       "      <th>ACH-001560</th>\n",
       "      <th>ACH-001561</th>\n",
       "      <th>ACH-001562</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>TSPAN6</th>\n",
       "      <td>4.447579</td>\n",
       "      <td>5.802452</td>\n",
       "      <td>4.794936</td>\n",
       "      <td>4.831371</td>\n",
       "      <td>6.491532</td>\n",
       "      <td>5.399855</td>\n",
       "      <td>5.391974</td>\n",
       "      <td>4.888013</td>\n",
       "      <td>4.731726</td>\n",
       "      <td>5.753818</td>\n",
       "      <td>...</td>\n",
       "      <td>3.060047</td>\n",
       "      <td>0.773996</td>\n",
       "      <td>2.769772</td>\n",
       "      <td>4.280956</td>\n",
       "      <td>4.089159</td>\n",
       "      <td>4.799087</td>\n",
       "      <td>4.275007</td>\n",
       "      <td>4.621173</td>\n",
       "      <td>4.399855</td>\n",
       "      <td>4.628774</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DPM1</th>\n",
       "      <td>7.227760</td>\n",
       "      <td>5.997744</td>\n",
       "      <td>6.929436</td>\n",
       "      <td>6.498570</td>\n",
       "      <td>6.672850</td>\n",
       "      <td>6.307246</td>\n",
       "      <td>6.866661</td>\n",
       "      <td>7.254840</td>\n",
       "      <td>5.849499</td>\n",
       "      <td>5.514438</td>\n",
       "      <td>...</td>\n",
       "      <td>6.837691</td>\n",
       "      <td>5.908333</td>\n",
       "      <td>7.010220</td>\n",
       "      <td>7.745170</td>\n",
       "      <td>7.149137</td>\n",
       "      <td>7.057992</td>\n",
       "      <td>7.379725</td>\n",
       "      <td>7.168923</td>\n",
       "      <td>7.191109</td>\n",
       "      <td>7.704941</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SCYL3</th>\n",
       "      <td>2.405992</td>\n",
       "      <td>1.970854</td>\n",
       "      <td>2.952334</td>\n",
       "      <td>2.414136</td>\n",
       "      <td>2.475085</td>\n",
       "      <td>2.017922</td>\n",
       "      <td>1.803227</td>\n",
       "      <td>2.289834</td>\n",
       "      <td>2.503349</td>\n",
       "      <td>2.536053</td>\n",
       "      <td>...</td>\n",
       "      <td>2.472488</td>\n",
       "      <td>1.831877</td>\n",
       "      <td>1.895303</td>\n",
       "      <td>2.304511</td>\n",
       "      <td>2.220330</td>\n",
       "      <td>2.204767</td>\n",
       "      <td>2.000000</td>\n",
       "      <td>1.871844</td>\n",
       "      <td>2.060047</td>\n",
       "      <td>2.589763</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C1orf112</th>\n",
       "      <td>4.594549</td>\n",
       "      <td>3.784504</td>\n",
       "      <td>3.709291</td>\n",
       "      <td>4.527946</td>\n",
       "      <td>4.486714</td>\n",
       "      <td>3.671293</td>\n",
       "      <td>3.841973</td>\n",
       "      <td>3.795975</td>\n",
       "      <td>3.761285</td>\n",
       "      <td>4.566206</td>\n",
       "      <td>...</td>\n",
       "      <td>4.571677</td>\n",
       "      <td>4.060912</td>\n",
       "      <td>1.550901</td>\n",
       "      <td>4.306700</td>\n",
       "      <td>4.551516</td>\n",
       "      <td>3.526069</td>\n",
       "      <td>4.251719</td>\n",
       "      <td>2.153805</td>\n",
       "      <td>3.231125</td>\n",
       "      <td>3.066950</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CFH</th>\n",
       "      <td>3.404631</td>\n",
       "      <td>2.750607</td>\n",
       "      <td>3.560715</td>\n",
       "      <td>2.153805</td>\n",
       "      <td>0.014355</td>\n",
       "      <td>1.214125</td>\n",
       "      <td>0.807355</td>\n",
       "      <td>0.367371</td>\n",
       "      <td>5.036503</td>\n",
       "      <td>3.881665</td>\n",
       "      <td>...</td>\n",
       "      <td>0.056584</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.704872</td>\n",
       "      <td>2.901108</td>\n",
       "      <td>5.655638</td>\n",
       "      <td>3.142413</td>\n",
       "      <td>3.300124</td>\n",
       "      <td>0.111031</td>\n",
       "      <td>2.857981</td>\n",
       "      <td>6.017254</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>UPK3BL2</th>\n",
       "      <td>3.279471</td>\n",
       "      <td>2.729009</td>\n",
       "      <td>3.976364</td>\n",
       "      <td>2.643856</td>\n",
       "      <td>5.422233</td>\n",
       "      <td>2.073820</td>\n",
       "      <td>3.065228</td>\n",
       "      <td>3.572890</td>\n",
       "      <td>3.241840</td>\n",
       "      <td>3.084064</td>\n",
       "      <td>...</td>\n",
       "      <td>2.353323</td>\n",
       "      <td>0.863938</td>\n",
       "      <td>0.565597</td>\n",
       "      <td>0.516015</td>\n",
       "      <td>0.650765</td>\n",
       "      <td>1.565597</td>\n",
       "      <td>0.163499</td>\n",
       "      <td>0.887525</td>\n",
       "      <td>0.137504</td>\n",
       "      <td>2.063503</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AC093512.2</th>\n",
       "      <td>2.653060</td>\n",
       "      <td>3.700440</td>\n",
       "      <td>2.214125</td>\n",
       "      <td>2.403268</td>\n",
       "      <td>2.933573</td>\n",
       "      <td>3.944858</td>\n",
       "      <td>2.965323</td>\n",
       "      <td>3.026800</td>\n",
       "      <td>2.778209</td>\n",
       "      <td>4.265287</td>\n",
       "      <td>...</td>\n",
       "      <td>4.765535</td>\n",
       "      <td>5.610877</td>\n",
       "      <td>2.424922</td>\n",
       "      <td>4.426265</td>\n",
       "      <td>2.324811</td>\n",
       "      <td>2.344828</td>\n",
       "      <td>1.799087</td>\n",
       "      <td>1.899176</td>\n",
       "      <td>2.904966</td>\n",
       "      <td>2.720278</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ARHGAP11B</th>\n",
       "      <td>1.827819</td>\n",
       "      <td>1.695994</td>\n",
       "      <td>0.903038</td>\n",
       "      <td>2.666757</td>\n",
       "      <td>2.195348</td>\n",
       "      <td>1.056584</td>\n",
       "      <td>1.655352</td>\n",
       "      <td>2.996389</td>\n",
       "      <td>1.682573</td>\n",
       "      <td>2.364572</td>\n",
       "      <td>...</td>\n",
       "      <td>1.807355</td>\n",
       "      <td>1.541019</td>\n",
       "      <td>0.613532</td>\n",
       "      <td>0.773996</td>\n",
       "      <td>1.176323</td>\n",
       "      <td>0.565597</td>\n",
       "      <td>1.744161</td>\n",
       "      <td>0.137504</td>\n",
       "      <td>0.956057</td>\n",
       "      <td>1.550901</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ABCF2-H2BE1</th>\n",
       "      <td>1.521051</td>\n",
       "      <td>3.132577</td>\n",
       "      <td>1.646163</td>\n",
       "      <td>0.855990</td>\n",
       "      <td>2.100978</td>\n",
       "      <td>2.797013</td>\n",
       "      <td>2.419539</td>\n",
       "      <td>0.344828</td>\n",
       "      <td>2.981853</td>\n",
       "      <td>2.482848</td>\n",
       "      <td>...</td>\n",
       "      <td>0.823749</td>\n",
       "      <td>0.189034</td>\n",
       "      <td>3.712596</td>\n",
       "      <td>1.007196</td>\n",
       "      <td>3.446256</td>\n",
       "      <td>2.592158</td>\n",
       "      <td>1.150560</td>\n",
       "      <td>2.754888</td>\n",
       "      <td>1.765535</td>\n",
       "      <td>3.129283</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>POLR2J3</th>\n",
       "      <td>4.516646</td>\n",
       "      <td>5.238023</td>\n",
       "      <td>5.458119</td>\n",
       "      <td>6.648465</td>\n",
       "      <td>5.239933</td>\n",
       "      <td>5.650190</td>\n",
       "      <td>6.129489</td>\n",
       "      <td>4.511595</td>\n",
       "      <td>5.559798</td>\n",
       "      <td>5.768449</td>\n",
       "      <td>...</td>\n",
       "      <td>5.689020</td>\n",
       "      <td>3.447579</td>\n",
       "      <td>5.375735</td>\n",
       "      <td>3.456806</td>\n",
       "      <td>4.397118</td>\n",
       "      <td>5.673556</td>\n",
       "      <td>3.759156</td>\n",
       "      <td>4.130931</td>\n",
       "      <td>4.688740</td>\n",
       "      <td>5.414474</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>11827 rows × 206 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             ACH-000010  ACH-000012  ACH-000015  ACH-000021  ACH-000029  \\\n",
       "gene                                                                      \n",
       "TSPAN6         4.447579    5.802452    4.794936    4.831371    6.491532   \n",
       "DPM1           7.227760    5.997744    6.929436    6.498570    6.672850   \n",
       "SCYL3          2.405992    1.970854    2.952334    2.414136    2.475085   \n",
       "C1orf112       4.594549    3.784504    3.709291    4.527946    4.486714   \n",
       "CFH            3.404631    2.750607    3.560715    2.153805    0.014355   \n",
       "...                 ...         ...         ...         ...         ...   \n",
       "UPK3BL2        3.279471    2.729009    3.976364    2.643856    5.422233   \n",
       "AC093512.2     2.653060    3.700440    2.214125    2.403268    2.933573   \n",
       "ARHGAP11B      1.827819    1.695994    0.903038    2.666757    2.195348   \n",
       "ABCF2-H2BE1    1.521051    3.132577    1.646163    0.855990    2.100978   \n",
       "POLR2J3        4.516646    5.238023    5.458119    6.648465    5.239933   \n",
       "\n",
       "             ACH-000030  ACH-000033  ACH-000035  ACH-000062  ACH-000066  ...  \\\n",
       "gene                                                                     ...   \n",
       "TSPAN6         5.399855    5.391974    4.888013    4.731726    5.753818  ...   \n",
       "DPM1           6.307246    6.866661    7.254840    5.849499    5.514438  ...   \n",
       "SCYL3          2.017922    1.803227    2.289834    2.503349    2.536053  ...   \n",
       "C1orf112       3.671293    3.841973    3.795975    3.761285    4.566206  ...   \n",
       "CFH            1.214125    0.807355    0.367371    5.036503    3.881665  ...   \n",
       "...                 ...         ...         ...         ...         ...  ...   \n",
       "UPK3BL2        2.073820    3.065228    3.572890    3.241840    3.084064  ...   \n",
       "AC093512.2     3.944858    2.965323    3.026800    2.778209    4.265287  ...   \n",
       "ARHGAP11B      1.056584    1.655352    2.996389    1.682573    2.364572  ...   \n",
       "ABCF2-H2BE1    2.797013    2.419539    0.344828    2.981853    2.482848  ...   \n",
       "POLR2J3        5.650190    6.129489    4.511595    5.559798    5.768449  ...   \n",
       "\n",
       "             ACH-001386  ACH-001549  ACH-001555  ACH-001556  ACH-001557  \\\n",
       "gene                                                                      \n",
       "TSPAN6         3.060047    0.773996    2.769772    4.280956    4.089159   \n",
       "DPM1           6.837691    5.908333    7.010220    7.745170    7.149137   \n",
       "SCYL3          2.472488    1.831877    1.895303    2.304511    2.220330   \n",
       "C1orf112       4.571677    4.060912    1.550901    4.306700    4.551516   \n",
       "CFH            0.056584    0.000000    1.704872    2.901108    5.655638   \n",
       "...                 ...         ...         ...         ...         ...   \n",
       "UPK3BL2        2.353323    0.863938    0.565597    0.516015    0.650765   \n",
       "AC093512.2     4.765535    5.610877    2.424922    4.426265    2.324811   \n",
       "ARHGAP11B      1.807355    1.541019    0.613532    0.773996    1.176323   \n",
       "ABCF2-H2BE1    0.823749    0.189034    3.712596    1.007196    3.446256   \n",
       "POLR2J3        5.689020    3.447579    5.375735    3.456806    4.397118   \n",
       "\n",
       "             ACH-001558  ACH-001559  ACH-001560  ACH-001561  ACH-001562  \n",
       "gene                                                                     \n",
       "TSPAN6         4.799087    4.275007    4.621173    4.399855    4.628774  \n",
       "DPM1           7.057992    7.379725    7.168923    7.191109    7.704941  \n",
       "SCYL3          2.204767    2.000000    1.871844    2.060047    2.589763  \n",
       "C1orf112       3.526069    4.251719    2.153805    3.231125    3.066950  \n",
       "CFH            3.142413    3.300124    0.111031    2.857981    6.017254  \n",
       "...                 ...         ...         ...         ...         ...  \n",
       "UPK3BL2        1.565597    0.163499    0.887525    0.137504    2.063503  \n",
       "AC093512.2     2.344828    1.799087    1.899176    2.904966    2.720278  \n",
       "ARHGAP11B      0.565597    1.744161    0.137504    0.956057    1.550901  \n",
       "ABCF2-H2BE1    2.592158    1.150560    2.754888    1.765535    3.129283  \n",
       "POLR2J3        5.673556    3.759156    4.130931    4.688740    5.414474  \n",
       "\n",
       "[11827 rows x 206 columns]"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "lung.expressed(threshold=0.50, style=\"values\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Cancer and CellLineCluster objects have an additional method that outputs a binary matrix\n",
    "of which genes/cell lines have mutations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>A1BG</th>\n",
       "      <th>A1CF</th>\n",
       "      <th>A2M</th>\n",
       "      <th>A2ML1</th>\n",
       "      <th>A3GALT2</th>\n",
       "      <th>A4GALT</th>\n",
       "      <th>A4GNT</th>\n",
       "      <th>AAAS</th>\n",
       "      <th>AACS</th>\n",
       "      <th>AADAC</th>\n",
       "      <th>...</th>\n",
       "      <th>ZWILCH</th>\n",
       "      <th>ZWINT</th>\n",
       "      <th>ZXDA</th>\n",
       "      <th>ZXDB</th>\n",
       "      <th>ZXDC</th>\n",
       "      <th>ZYG11A</th>\n",
       "      <th>ZYG11B</th>\n",
       "      <th>ZYX</th>\n",
       "      <th>ZZEF1</th>\n",
       "      <th>ZZZ3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ACH-000523</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACH-000749</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACH-000787</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACH-000852</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACH-000867</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACH-000521</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACH-000010</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACH-000589</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACH-000575</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACH-000587</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>273 rows × 17376 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            A1BG  A1CF  A2M  A2ML1  A3GALT2  A4GALT  A4GNT  AAAS  AACS  AADAC  \\\n",
       "ACH-000523     1     0    0      0        0       0      0     0     0      0   \n",
       "ACH-000749     1     0    0      0        0       0      0     0     0      0   \n",
       "ACH-000787     1     1    0      0        0       0      0     0     0      0   \n",
       "ACH-000852     1     0    1      1        0       0      0     0     0      0   \n",
       "ACH-000867     1     0    0      0        0       0      0     0     0      0   \n",
       "...          ...   ...  ...    ...      ...     ...    ...   ...   ...    ...   \n",
       "ACH-000521     0     0    0      0        0       0      0     0     0      0   \n",
       "ACH-000010     0     0    0      0        0       0      0     0     0      0   \n",
       "ACH-000589     0     0    0      0        0       0      0     0     0      0   \n",
       "ACH-000575     0     0    0      0        0       0      0     0     0      0   \n",
       "ACH-000587     0     0    0      0        0       0      0     0     0      0   \n",
       "\n",
       "            ...  ZWILCH  ZWINT  ZXDA  ZXDB  ZXDC  ZYG11A  ZYG11B  ZYX  ZZEF1  \\\n",
       "ACH-000523  ...       0      1     0     0     0       0       0    0      0   \n",
       "ACH-000749  ...       0      0     0     0     0       0       0    0      0   \n",
       "ACH-000787  ...       0      0     0     0     0       0       0    0      0   \n",
       "ACH-000852  ...       0      0     0     0     0       0       1    0      0   \n",
       "ACH-000867  ...       1      0     0     0     0       0       0    0      0   \n",
       "...         ...     ...    ...   ...   ...   ...     ...     ...  ...    ...   \n",
       "ACH-000521  ...       0      0     0     0     0       0       0    0      0   \n",
       "ACH-000010  ...       0      0     0     0     0       0       0    0      0   \n",
       "ACH-000589  ...       0      0     0     0     0       0       0    0      0   \n",
       "ACH-000575  ...       0      0     0     0     0       0       0    0      0   \n",
       "ACH-000587  ...       0      0     0     0     0       0       0    0      0   \n",
       "\n",
       "            ZZZ3  \n",
       "ACH-000523     0  \n",
       "ACH-000749     0  \n",
       "ACH-000787     0  \n",
       "ACH-000852     0  \n",
       "ACH-000867     0  \n",
       "...          ...  \n",
       "ACH-000521     0  \n",
       "ACH-000010     0  \n",
       "ACH-000589     0  \n",
       "ACH-000575     0  \n",
       "ACH-000587     0  \n",
       "\n",
       "[273 rows x 17376 columns]"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "lung.mutation_matrix()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Organelle Methods and Attributes\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Attributes:\n",
      "\n",
      "conf: 3\n",
      "get_name: Plasma membrane\n",
      "location: Plasma membrane\n",
      "genes list first item: ABCA7\n",
      "genes length: 1547\n",
      "\n",
      "Methods:\n",
      "\n",
      "cn_normal\n",
      "deletion\n",
      "dependency_of\n",
      "dependent\n",
      "duplication\n",
      "effect_of\n",
      "essential\n",
      "expressed\n",
      "expression_of\n",
      "genes_and_conf\n",
      "mutated\n",
      "non_dependent\n",
      "non_essential\n",
      "unexpressed\n"
     ]
    }
   ],
   "source": [
    "pretty_print_attr(membrane)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
